pvlib/irradiance.py

```diff
@@ -1945,3 +1945,106 @@ def _get_dirint_coeffs():
 
     return coeffs[1:, 1:, :, :]
+
+
+def dni(ghi, dhi, zenith, method='clearsky', clearsky_dni=None,
+        set_to_nan=False, clearsky_tolerance=1, **kwargs):
+    """
+    Determine DNI from GHI and DHI.
+
+    When calculating the DNI from GHI and DHI the calculated DNI may be
+    unreasonably high or negative for zenith angles close to 90 degrees
+    (sunrise/sunset transitions). This function offers different methods
+    to correct those values.
+
+    Parameters
+    ----------
+    ghi : Series
+        Global horizontal irradiance.
+
+    dhi : Series
+        Diffuse horizontal irradiance.
+
+    zenith : Series
+        True (not refraction-corrected) zenith angles in decimal
+        degrees. Angles must be >=0 and <=180.
+
+    method : str
+        The method used to correct the calculated DNI.
+        Must be one of 'clearsky', 'cutoff'.
+    
+    clearsky_dni : None or Series
+        Clearsky direct normal irradiance.
+    
+    set_to_nan : boolean
+        If True the values of the calculated DNI that need correction will be
+        replaced by NaN. Otherwise they will be replaced by the calculated 
+        value depending on the specified method.
+
+    clearsky_tolerance : float
+        If method 'clearsky' is chosen this parameter can be used to allow a
+        tolerance by how much the calculated DNI value can be greater than
+        the clearsky value before it is identified as an unreasonable value.
+
+    Returns
+    -------
+    dni : Series
+        The modeled direct normal irradiance.
+
+    """
+
+    # calculate DNI
+    dni_tmp = (ghi - dhi) / tools.cosd(zenith)
+    dni = dni_tmp.copy()
+    
+    # cutoff negative values
+    if set_to_nan:
+        dni[dni < 0] = float('nan')
+    else:
+        dni[dni < 0] = 0
+    
+    # correct unreasonable values
+    if method == 'clearsky':
+        if set_to_nan:
+            # set non-zero DNI values for zenith angles >= 90.0 to NaN
+            dni[(zenith >= 90.0) & (dni != 0)] = float('nan')
+            # set DNI values for zenith angles between 80.0 and 90.0 degrees
+            # that are greater than the calculated clearsky DNI to NaN
+            dni[(zenith >= 80.0) & (zenith <= 90.0) & (dni > clearsky_dni)] = \
+                float('nan')
+        else:
+            # set DNI for zenith angles >= 90.0 degrees to zero
+            dni[(zenith >= 90.0)] = 0
+            # cut DNI for zenith angles between 80.0 and 90.0 degrees to
+            # maximum value given by the clearsky DNI
+            dni[(zenith >= 80.0) & (zenith < 90.0) &
+                (dni > (clearsky_dni * clearsky_tolerance))] = \
+                    (clearsky_dni * clearsky_tolerance)
+    elif method == 'cutoff':
+        if set_to_nan:
+            dni[(zenith > 88) & (dni != 0)] = float('nan')
+        else:
+            dni[zenith > 88] = 0
+    else:
+        raise ValueError('Invalid method: %s', method)
+
+    # debugging info how many values were corrected
```

Comment:
Sure, I can also take it out if you like. I think it is a nice debugging feature, though. Maybe I could reduce it down to 2-3 lines?